pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BaseTableUpsertMetadataManager.java

```diff
@@ -75,6 +115,145 @@ public void init(TableConfig tableConfig, Schema schema, TableDataManager tableD
 
     _enableSnapshot = upsertConfig.isEnableSnapshot();
     _serverMetrics = serverMetrics;
+    _helixManager = helixManager;
+    _preloadExecutor = preloadExecutor;
+    try {
+      if (_enableSnapshot && upsertConfig.isEnablePreload()) {
+        preloadSegments();
```

Comment:
(MAJOR) Currently we are blocking the `init()` to preload all the segments, and this will prevent table data manager from being created. I feel it might work, but it is not intended. With the current code, we don't even need the lock because before preloading is done, table data manager won't be created.

Reply:
Good point! The concurrentMap.computeIfAbsent() used to create table data mgr conditionally in InstanceDataMgr is functioning as the lock. This actually reminded me of the logs I observed in my tests.

Reply:
Also the table data manager is created when first message is received in the Helix queue iiuc. We might need to change that as well.

Reply:
PR updated. The explicit waiting logic with lock/condition was removed, replaced with the implicit waiting logic as ensured by ConcurrentHashMap.computeIfAbsent() in HelixInstanceDataManager. I left a comment about this implicit behavior.

hey @KKcorps 
> Also the table data manager is created when first message is received in the Helix queue iiuc.
Yes. This is the existing implementation of HelixInstanceDataManager, e.g. 
```
public void addRealtimeSegment(String realtimeTableName, String segmentName)
...
_tableDataManagerMap.computeIfAbsent(realtimeTableName, k -> createTableDataManager(k, tableConfig))
...
```

> We might need to change that as well.  
Pardon me, I didn't get where to change. Could you help elaborate a bit more on this. Thank you.

Reply:
I meant was we will need to create table data manager before we even received the first helix message right?
Otherwise the preloading will trigger at the exact same time as we start loading the segments.

Reply:
In fact, the `createTableDataManager()` method pasted above creates the TableDataManager object first, and then calls its init() method. The preloading logic happens as part of the init() method. So among the Helix threads about to handle segment state transitions, only the first Helix thread would proceed and preload segments, and the other Helix threads would wait for it to finish, as ensured by ConcurrentHashMap.computeIfAbsent(... lambda), which runs the lambda atomically